### DISCUSSION TITLE
Change to value of Property in ViewModel is not displayed in bound Label on UI

### DISCUSSION THREAD

**Discussion by tradegear** [CommunityToolkit/dotnet#431] [Q&A] [answered]
Trying to do something that would seem basic.  I'm probably missing an attribute setting.
The issue is that in code I change the value of a property within a ViewModel, and I'd expect that to be reflect within the UI.  It does not.  I'm clearly starting from the basics here.  I'm starting from the bootstrap bot project in a fresh Maui app, then piecing extracts and snippets from hours of Montemagno's youtube videos and code samples.  I've tried to copy the pertinent code snippets and removed unimportant stuff.

MainPage.xaml:
-------------------------------------------------------
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MTRMauiClient.MainPage"
             xmlns:viewmodel="clr-namespace:MTRMauiClient.ViewModels"
             x:DataType="viewmodel:SessionTokenViewModel">

    <ScrollView>
        <VerticalStackLayout
            Spacing="25"
            Padding="30,0"
            VerticalOptions="Center">

            <Entry Placeholder="Enter Device Key"
		        Text="{Binding DeviceKey}"
		        HorizontalOptions="Center"/>

            <Entry Placeholder="Enter Password"
		        Text="{Binding Password}"
		        HorizontalOptions="Center"/>

            <Button Text="Get Session Token"
	            Command="{Binding GetCommand}"
	            HorizontalOptions="Center"/>

            <Label
                Text="{Binding DeviceKey}"
                SemanticProperties.HeadingLevel="Level1"
                FontSize="20"
                HorizontalOptions="Center" />

        </VerticalStackLayout>
    </ScrollView>
</ContentPage>
-----------------------------------------------

MainPage.xaml.cs:
------------------------------------------
namespace MTRMauiClient;

public partial class MainPage : ContentPage
{
	//int count = 0;

	public MainPage(SessionTokenViewModel sessionTokenViewModel)
	{
		InitializeComponent();
		BindingContext = sessionTokenViewModel;
	}
}
--------------------------------------------------
MauiProgram.cs:
----------------------------------------------------
using MTRMauiClient.ViewModels;
namespace MTRMauiClient;

public static class MauiProgram
{
	public static MauiApp CreateMauiApp()
	{
		var builder = MauiApp.CreateBuilder();
		builder
			.UseMauiApp<App>()
			.ConfigureFonts(fonts =>
			{
				fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
				fonts.AddFont("OpenSans-Semibold.ttf", "OpenSansSemibold");
			});

		builder.Services.AddSingleton<MainPage>();
		builder.Services.AddSingleton<SessionTokenViewModel>();
		return builder.Build();
	}
}
--------------------------------------------------------
SessionTokenViewModel.cs:
-----------------------------------------------------
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace MTRMauiClient.ViewModels
{
    public partial class SessionTokenViewModel : ObservableObject  /* implements INotifyPropertyChanged automatically and lots of other stuff */
    {
        [ObservableProperty]
        private string deviceKey;
        [ObservableProperty]
        private string password;
 
        [RelayCommand]
        void Get()
        {
            deviceKey = $"!{deviceKey}!"; 
/* just modify to anything different to prove the ViewModel value changed, so the UI can be validated */
            string x = deviceKey; 
/* for debugging to make sure deviceKey truly was updated wihtin the ViewModel, and I have confirmed it IS. 
Why does this change to this value NOT get reflected within the UI Label where Text="{Binding DeviceKey} ?!?*/
        }
    }
}
-------------------------------------------------

**Question:**
Why is it that when this code in "Get()" changes the value for deviceKey, the UI is not reflecting the data change?
At first I suspected an Entry Placeholder might not accept an update from code, so I created the 
<Label
                Text="{Binding Password}"
....
to prove that out.  This also does NOT reflect the value change to deviceKey

Thanks!

**Comment by powerdude** [comment, marked as answer]
you're changing the field, not the property that will raise the event(s).  change the code to:
```csharp
  DeviceKey = $"!{deviceKey}!";
```

**Reply by tradegear** [reply]
Thank you! - and may all our problems be this simple.  
Between when I posted and your answer, I had also tried: 
Text="{Binding DeviceKey, Mode=TwoWay}" 
Which (a) didn't solve the problem, and (b) I've just tested - wasn't necessary due to the types of controls I'd guess.  But probably something to keep in mind for later.
